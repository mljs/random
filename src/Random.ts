import { XSadd } from 'ml-xsadd';

import { ChoiceOptions } from './Options';
import { RandomNumberGenerator } from './RandomNumberGenerator';
import choice from './choice';

/**
 * @classdesc Random class
 */
export class Random {
  private randomGenerator: RandomNumberGenerator;
  /**
   * @param [seedOrRandom=Math.random] - Control the random number generator used by the Random class instance. Pass a random number generator function with a uniform distribution over the half-open interval [0, 1[. If seed will pass it to ml-xsadd to create a seeded random number generator. If undefined will use Math.random.
   */
  constructor(seedOrRandom: RandomNumberGenerator | number = Math.random) {
    if (typeof seedOrRandom === 'number') {
      const xsadd = new XSadd(seedOrRandom);
      this.randomGenerator = xsadd.random;
    } else {
      this.randomGenerator = seedOrRandom;
    }
  }

  /**
   * Returns an array of elements choosen from a list
   * @param values  - The values to choose from. If a number, the list will be a range of integer from 0 to that number.
   * @param options - option object
   * @returns The choosen values
   */

  public choice<T>(values: T[], options?: ChoiceOptions): T[];
  public choice(values: number, options?: ChoiceOptions): number[];
  public choice<T>(
    values: T[] | number,
    options?: ChoiceOptions,
  ): Array<T | number> {
    if (typeof values === 'number') {
      return choice(values, options, this.randomGenerator);
    }
    return choice(values, options, this.randomGenerator);
  }

  /**
   * Draw a random number from a uniform distribution on [0,1)
   * @returns The random number
   */
  public random(): number {
    return this.randomGenerator();
  }

  /**
   * Draw a random integer from a uniform distribution on [low, high). If only low is specified, the number is drawn on [0, low)
   * @param low - The lower bound of the uniform distribution interval.
   * @param high - The higher bound of the uniform distribution interval.
   */
  public randInt(low: number, high?: number): number {
    if (high === undefined) {
      high = low;
      low = 0;
    }
    return low + Math.floor(this.randomGenerator() * (high - low));
  }

  /**
   * Draw several random number from a uniform distribution on [0, 1)
   * @param size - The number of number to draw
   * @returns - The list of drawn numbers.
   */
  public randomSample(size: number): number[] {
    const result: number[] = [];
    for (let i = 0; i < size; i++) {
      result.push(this.random());
    }
    return result;
  }
}
