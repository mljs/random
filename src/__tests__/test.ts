import XSAdd from 'ml-xsadd';
import Random from '..';

let random: Random;
describe('random.choice', () => {
  beforeEach(() => {
    random = new Random(28);
  });
  it('should return all elements', () => {
    random = new Random();
    expect(random.choice(13, { size: 13 }).sort((a, b) => a - b)).toEqual([
      0,
      1,
      2,
      3,
      4,
      5,
      6,
      7,
      8,
      9,
      10,
      11,
      12
    ]);
  });

  it('should throw because size option is too large', () => {
    expect(() => {
      random.choice(2, { size: 3 });
    }).toThrow(/too large/);
  });

  it('should select elements without replacement', () => {
    const xsadd = new XSAdd(28);
    random = new Random(xsadd.random);
    expect(random.choice(10, { size: 10 })).toMatchSnapshot();
  });

  it('should choose values with replacement', () => {
    expect(random.choice(10, { size: 20, replace: true })).toMatchSnapshot();
  });

  it('pass values as an array', () => {
    expect(
      random
        .choice([1, 2, 3, 4, 5, 6, 7, 8, 9], {
          size: 9
        })
        .sort()
    ).toEqual([1, 2, 3, 4, 5, 6, 7, 8, 9]);
  });

  it('should produce choice given probabilities', () => {
    const samples = 10000;
    const r = random.choice(2, {
      size: samples,
      replace: true,
      probabilities: [0.3, 0.7]
    });
    const count = r.reduce(
      (prev, current) => (current === 0 ? prev + 1 : prev),
      0
    );
    expect(count / samples).toBeCloseTo(0.3, 2);
  });

  it('should throw if probabilities option has the wrong size', () => {
    expect(() => {
      random.choice(3, { replace: true, probabilities: [0.5, 0.5] });
    }).toThrow(
      'the length of probabilities option should be equal to the number of choices'
    );
  });

  it('should throw if probabilities option does not sum to 1', () => {
    expect(() => {
      random.choice(3, { replace: true, probabilities: [0.1, 0.5, 0.4001] });
    }).toThrow('probabilities should sum to 1');
  });

  it('choice with probabilities and no replacement not implemented', () => {
    expect(() => {
      random.choice(3, { replace: false, probabilities: [0.1, 0.5, 0.4] });
    }).toThrow(/not implemented/);
  });
});

describe('random.random', () => {
  it('should generate random numbers', () => {
    random = new Random(14);
    const r = [];
    for (let i = 0; i < 10; i++) {
      r.push(random.random());
    }
    expect(r).toMatchSnapshot();
  });
});

describe('random.randomSample', () => {
  it('should generate an array of random numbers', () => {
    const numbers = random.randomSample(10);
    expect(numbers).toHaveLength(10);
    numbers.forEach(num => expect(num).toBeLessThan(1));
    numbers.forEach(num => expect(num).toBeGreaterThanOrEqual(0));
    expect(numbers).toMatchSnapshot();
  });
});

describe('random.randInt', () => {
  it('should generate integers within an range with uniform distribution', () => {
    const n = 50000;
    const low = 4;
    const high = 8;
    const values = [];
    for (let i = 0; i < n; i++) {
      values.push(random.randInt(low, high));
    }
    const expFreq = 1 / (high - low);
    expect(freq(values, 4)).toBeCloseTo(expFreq);
    expect(freq(values, 5)).toBeCloseTo(expFreq);
    expect(freq(values, 6)).toBeCloseTo(expFreq);
    expect(freq(values, 7)).toBeCloseTo(expFreq);

    expect(freq(values, 3)).toEqual(0);
    expect(freq(values, 8)).toEqual(0);
  });

  it('if only low is specified, low is 0 and high is low', () => {
    const rand1 = new Random(12);
    const rand2 = new Random(12);
    const n = 20;
    const values1 = [];
    const values2 = [];
    for (let i = 0; i < n; i++) {
      values1.push(rand1.randInt(2));
      values2.push(rand2.randInt(0, 2));
    }
    expect(values1).toEqual(values2);
  });
});

function freq(arr: number[], n: number) {
  return (
    arr.reduce((prev, current) => (current === n ? prev + 1 : prev), 0) /
    arr.length
  );
}
