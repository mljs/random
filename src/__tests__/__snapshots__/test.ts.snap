// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`random.choice should choose values with replacement 1`] = `
Array [
  4,
  5,
  3,
  9,
  0,
  7,
  2,
  3,
  2,
  3,
  0,
  9,
  3,
  8,
  6,
  5,
  7,
  2,
  5,
  5,
]
`;

exports[`random.choice should select elements without replacement 1`] = `
Array [
  4,
  5,
  3,
  9,
  0,
  7,
  1,
  2,
  6,
  8,
]
`;

exports[`random.random should generate random numbers 1`] = `
Array [
  0.36992722749710083,
  0.009614884853363037,
  0.11960279941558838,
  0.541417121887207,
  0.07362216711044312,
  0.20413708686828613,
  0.36353737115859985,
  0.18186044692993164,
  0.9514840841293335,
  0.7254511713981628,
]
`;

exports[`random.randomSample should generate an array of random numbers 1`] = `
Array [
  0.7623295187950134,
  0.24819517135620117,
  0.6318815350532532,
  0.5753740072250366,
  0.860216498374939,
  0.2248849868774414,
  0.8180570006370544,
  0.8344047665596008,
  0.8892887234687805,
  0.47667670249938965,
]
`;
